docs/source/advanced.rst

```diff
@@ -179,29 +196,95 @@ Custom components can be registered via the following modification of the ``entr
        ],
    }
 
-The line above registers a group ``foo`` that is associated with the module ``my_project.bar``.
-TorchX will recursively traverse lowest level dir associated with the ``my_project.bar`` and will find
-all defined components.
-
-.. note:: If there are two registry entries, e.g. ``foo = my_project.bar`` and ``test = my_project``
-          there will be two sets of overlapping components with different aliases.
+TorchX will search the module ``my_project.bar`` for all defined components and group the found
+components under the ``foo.*`` prefix. In this case, the component ``my_project.bar.baz.trainer``
+would be registered with the name ``foo.baz.trainer``.
 
+.. note::
+    Only python packages (those directories with an ``__init__.py`` file)
+    are searched for and TorchX makes no attempt to recurse into namespace packages
+    (directories without a ``__init__.py`` file).
+    However you may register a top level namespace package.
 
-After registration, torchx cli will display registered components via:
+``torchx`` CLI will display registered components via:
 
 .. code-block:: shell-session
 
  $ torchx builtins
+ Found 1 builtin components:
+ 1. foo.baz.trainer
 
-If ``my_project.bar`` had the following directory structure:
+The custom component can then be used as:
 
-::
+.. code-block:: shell-session
 
- $PROJECT_ROOT/my_project/bar/
-     |- baz.py
+ $ torchx run foo.baz.trainer -- --name "test app"
+
+
+When you register your own components, TorchX will not include its own builtins. To add TorchX's
+builtin components you must specify another entry as:
```

Comment:
may want to mention you can do this in the config as well?

Reply:
the one supported by `.torchxconfig` is super weird (I want to deprecate it in favor of supporting "alias" -> "module_name" style custom components from `.torchxconfig`). Currently what you can do is to have a section like

```
# .torchxconfig
[torchx.file]
get_file_contents = my.custom.file_reader:read
```

And when you run

```
torchx run my/custom.py:component_fn
```

We call `my.custom.file_reader.read("my/custom.py")` which we expect to load the `component_fn` into the global namespace. Took me a while to wrap my head around it, so I don't expect anyone to actually use it.

Instead I'd like to support something like this from `.torchxconfig`:

```
# .torchxconfig
[torchx.components]
foo = my.custom.component
_ = torchx.components
```

which would be the same as doing

```
# setup.py

entrypoints = {
  "torchx.components": [
        "foo = my.custom.component",
        "_ = torchx.components",
   ],
}
```